RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+        if (pfClusteringVars[rhIdx].pfrh_isSeed() && topoId > -1) {
+          // Allot the total number of rechits for this topo cluster for rh fractions
+          int offset = alpaka::atomicAdd(acc, &totalSeedFracOffset, pfClusteringVars[topoId].topoRHCount());
+
+          // Add offset for this PF cluster seed
+          pfClusteringVars[rhIdx].seedFracOffsets() = offset;
+
+          // Store recHitFraction offset & size information for each seed
+          clusterView[pfClusteringVars[rhIdx].rhIdxToSeedIdx()].rhfracOffset() =
+              pfClusteringVars[rhIdx].seedFracOffsets();
+          clusterView[pfClusteringVars[rhIdx].rhIdxToSeedIdx()].rhfracSize() =
+              pfClusteringVars[topoId].topoRHCount() - pfClusteringVars[topoId].topoSeedCount() + 1;
+        }
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if (once_per_block(acc)) {
+        pfClusteringVars.pcrhFracSize() = totalSeedFracOffset;
+        pfClusteringVars.nRHFracs() = totalSeedFracOffset;
+        clusterView.nRHFracs() = totalSeedFracOffset;
+        clusterView.nSeeds() = *nSeeds;
+        clusterView.nTopos() = pfClusteringVars.nTopos();
+
+        for (int i = 0; i < nRH; i++) {
+          clusterView[i].topoRHCount() = pfClusteringVars[i].topoRHCount();
+        }
+        if (pfClusteringVars.pcrhFracSize() > 200000)  // Warning in case the fraction is too large
+          printf("At the end of topoClusterContraction, found large *pcrhFracSize = %d\n",
+                 pfClusteringVars.pcrhFracSize());
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+    }
+  };
+
+  // Prefill the rechit index for all PFCluster fractions
+  // Optimized for GPU parallel, but works on any backend
+  class FillRhfIndex {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFRecHitFractionDeviceCollection::View fracView) const {
+      const int nRH = pfRecHits.size();
+      int i = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] +
+              alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u] *
+                  alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];  // i is the seed index
+      int j = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[1u] +
+              alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[1u] *
+                  alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[1u];  // j is NOT a seed
+
+      if (i < nRH && j < nRH) {
+        int topoId = pfClusteringVars[i].pfrh_topoId();
+        if (topoId == pfClusteringVars[j].pfrh_topoId() && topoId > -1 && pfClusteringVars[i].pfrh_isSeed()) {
```

Comment:
if you reorder the checks with those based on `pfClusteringVars[i]` first, when they fail `pfClusteringVars[j]` will not be read at all:
```suggestion
        if (topoId > -1 && pfClusteringVars[i].pfrh_isSeed() && topoId == pfClusteringVars[j].pfrh_topoId()) {
```